// material-time-control/src/w-mat-timepicker/w-mat-timepicker.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var material_1 = require("@angular/material");
var w_time_dialog_component_1 = require("../w-time-dialog/w-time-dialog.component");
var WMatTimePickerComponent = /** @class */ (function () {
    function WMatTimePickerComponent(dialog) {
        this.dialog = dialog;
        this.disabled = false;
        this.userTimeChange = new core_1.EventEmitter();
    }
    WMatTimePickerComponent.prototype.ngOnInit = function () {
        if (!this.userTime) {
            this.userTime = {
                hour: 10,
                minute: 25,
                meriden: 'PM',
                format: 24
            };
        }
    };
    Object.defineProperty(WMatTimePickerComponent.prototype, "time", {
        get: function () {
            if (!this.userTime || this.disabled) {
                return '';
            }
            var meriden = "" + this.userTime.meriden;
            var hour = "" + this.userTime.hour;
            if (this.userTime.hour === 24) {
                hour = '00';
            }
            var tt = this.userTime.minute < 10 ? hour + ":0" + String(this.userTime.minute) : hour + ":" + this.userTime.minute;
            return this.userTime.format === 24 ? tt : tt + (" " + meriden);
        },
        enumerable: true,
        configurable: true
    });
    WMatTimePickerComponent.prototype.inputChange = function (ev) {
        var val = ev.target.value;
        console.log('input change : ', val);
        var _a = val.split(':'), hourStr = _a[0], minuteStr = _a[1];
        var hour = hourStr ? parseInt(hourStr, 10) : 0;
        if (this.userTime.format === 12) {
            var meridenPos = minuteStr.indexOf(' ');
            if (meridenPos > -1) {
                minuteStr = minuteStr.substr(0, meridenPos);
                var meriden = minuteStr.substr(meridenPos).toUpperCase();
                this.userTime.meriden = ['PM', 'AM'].indexOf(meriden) > -1 ? meriden : this.userTime.meriden;
            }
        }
        var minute = minuteStr ? parseInt(minuteStr, 10) : 0;
        console.log('minute, hour : ', minute, hour);
        var minHour = this.minTime ? this.minTime.hour : 0;
        var maxHour = this.maxTime ? this.maxTime.hour : this.userTime.format - 1;
        console.log('minHour, maxHour : ', minHour, maxHour);
        if (hour < minHour) {
            hour = minHour;
            minute = 0;
        }
        if (hour > maxHour) {
            hour = maxHour;
            minute = 59;
        }
        var minMinute = this.minTime && hour === minHour ? this.minTime.minute : 0;
        var maxMinute = this.maxTime && hour === maxHour ? this.maxTime.minute : 59;
        console.log('minMinute, maxMinute: ', minMinute, maxMinute);
        if (minute < minMinute) {
            minute = minMinute;
        }
        if (minute > maxMinute) {
            minute = maxMinute;
        }
        console.log('minute, hour : ', minute, hour);
        this.userTime.hour = hour;
        this.userTime.minute = minute;
        ev.target.value = this.time;
        this.emituserTimeChange();
    };
    WMatTimePickerComponent.prototype.showPicker = function ($event) {
        var _this = this;
        var dialogRef = this.dialog.open(w_time_dialog_component_1.WTimeDialogComponent, {
            data: {
                time: {
                    hour: this.userTime.hour,
                    minute: this.userTime.minute,
                    meriden: this.userTime.meriden,
                    format: this.userTime.format
                },
                minTime: this.minTime,
                maxTime: this.maxTime,
                color: this.color
            },
            width: '600px',
        });
        dialogRef.afterClosed()
            .subscribe(function (result) {
            // result will be update userTime object or -1 or undefined (closed dialog w/o clicking cancel)
            if (result === undefined || (typeof result === 'number' && result === -1)) {
                return;
            }
            else {
                _this.userTime = result;
                _this.emituserTimeChange();
            }
        });
        return false;
    };
    WMatTimePickerComponent.prototype.emituserTimeChange = function () {
        this.userTimeChange.emit(this.userTime);
    };
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object)
    ], WMatTimePickerComponent.prototype, "disabled", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object)
    ], WMatTimePickerComponent.prototype, "userTime", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object)
    ], WMatTimePickerComponent.prototype, "minTime", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object)
    ], WMatTimePickerComponent.prototype, "maxTime", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], WMatTimePickerComponent.prototype, "userTimeChange", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], WMatTimePickerComponent.prototype, "color", void 0);
    WMatTimePickerComponent = __decorate([
        core_1.Component({
            selector: 'w-mat-timepicker',
            styleUrls: ['./w-mat-timepicker.component.scss'],
            templateUrl: './w-mat-timepicker.component.html'
        }),
        __metadata("design:paramtypes", [material_1.MatDialog])
    ], WMatTimePickerComponent);
    return WMatTimePickerComponent;
}());
exports.WMatTimePickerComponent = WMatTimePickerComponent;
