// material-time-control/src/w-clock/w-clock.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
// Current type to show
var CLOCK_TYPE;
(function (CLOCK_TYPE) {
    CLOCK_TYPE[CLOCK_TYPE["HOURS"] = 1] = "HOURS";
    CLOCK_TYPE[CLOCK_TYPE["MINUTES"] = 2] = "MINUTES";
})(CLOCK_TYPE = exports.CLOCK_TYPE || (exports.CLOCK_TYPE = {}));
var WClockComponent = /** @class */ (function () {
    function WClockComponent() {
        this.userTimeChange = new core_1.EventEmitter();
        this.viewChange = new core_1.EventEmitter();
        this.steps = new Array();
    }
    WClockComponent.prototype.ngOnChanges = function () {
        this.setupUI();
    };
    WClockComponent.prototype.setupUI = function () {
        this.steps = new Array();
        switch (this.currentView) {
            case CLOCK_TYPE.HOURS:
                for (var i = 1; i <= this.userTime.format; i++) {
                    this.steps.push(i);
                    this.selectedTimePart = this.userTime.hour || 0;
                    if (this.selectedTimePart > this.userTime.format) {
                        this.selectedTimePart -= this.userTime.format;
                    }
                }
                break;
            case CLOCK_TYPE.MINUTES:
                for (var i = 1; i <= 59; i += 1) {
                    this.steps.push(i);
                }
                this.steps.push(0);
                this.selectedTimePart = this.userTime.minute || 0;
                break;
        }
    };
    WClockComponent.prototype.getPointerStyle = function () {
        var divider = 1;
        switch (this.currentView) {
            case CLOCK_TYPE.HOURS:
                divider = 12; // this.userTime.format;
                break;
            case CLOCK_TYPE.MINUTES:
                divider = 60;
                break;
        }
        var degrees = 0;
        if (this.currentView === CLOCK_TYPE.HOURS) {
            degrees = Math.round(this.userTime.hour * (360 / divider)) - 180;
        }
        else {
            degrees = Math.round(this.userTime.minute * (360 / divider)) - 180;
        }
        var style = {
            '-webkit-transform': 'rotate(' + degrees + 'deg)',
            '-ms-transform': 'rotate(' + degrees + 'deg)',
            'transform': 'rotate(' + degrees + 'deg)'
        };
        if (this.currentView === CLOCK_TYPE.HOURS && this.userTime.hour <= 12) {
            style = Object.assign(style, { height: '30%', bottom: '20%' });
        }
        return style;
    };
    WClockComponent.prototype.disabledTimeValue = function (step) {
        if (!this.minTime && !this.maxTime) {
            return false;
        }
        if (this.currentView === CLOCK_TYPE.HOURS) {
            if (this.minTime && this.minTime.hour > step) {
                return true;
            }
            return this.maxTime && this.maxTime.hour < step;
        }
        if (this.currentView === CLOCK_TYPE.MINUTES) {
            if ((this.minTime && this.minTime.hour > this.userTime.hour) || (this.maxTime && this.maxTime.hour < this.userTime.hour)) {
                return true;
            }
            if (this.minTime && this.maxTime && this.minTime.hour === this.userTime.hour && this.maxTime.hour === this.userTime.hour) {
                return step < this.minTime.minute || step > this.maxTime.minute;
            }
            if (this.minTime && this.minTime.hour === this.userTime.hour) {
                return step < this.minTime.minute;
            }
            if (this.maxTime && this.maxTime.hour === this.userTime.hour) {
                return step > this.maxTime.minute;
            }
            return false;
        }
    };
    WClockComponent.prototype.getTimeValueClass = function (step, index) {
        // const stepDeg = this.currentView === CLOCK_TYPE.HOURS ? 360 / this.userTime.format : 360 / 12;
        var stepDeg = this.currentView === CLOCK_TYPE.HOURS ? 360 / 12 : 360 / 60;
        // const stepDeg = this.currentView === CLOCK_TYPE.HOURS ? 360 / 12 : 360 / 12;
        var degrees = this.currentView === CLOCK_TYPE.HOURS && step > 12 ? (stepDeg * (index + 1 - 12)) : (stepDeg * (index + 1));
        var classes = 'w-clock-step w-clock-deg' + degrees;
        if (this.currentView === CLOCK_TYPE.MINUTES)
            classes += ' smaller';
        if (this.currentView === CLOCK_TYPE.HOURS && this.userTime.format === 24 && step <= 12)
            classes += ' inside';
        if (this.selectedTimePart === step) {
            classes += ' mat-primary';
        }
        return classes;
    };
    WClockComponent.prototype.changeTimeValue = function (step) {
        if (this.currentView === CLOCK_TYPE.HOURS) {
            this.userTime.hour = step;
            // auto switch to minutes
            this.viewChange.emit(CLOCK_TYPE.MINUTES);
        }
        else {
            this.userTime.minute = step;
            // auto switch to hours
            // this.viewChange.emit(CLOCK_TYPE.HOURS);
        }
        this.userTimeChange.emit(this.userTime);
    };
    WClockComponent.prototype.displayStep = function (step) {
        if (this.currentView === CLOCK_TYPE.HOURS && step === 24)
            return '00';
        if (this.currentView === CLOCK_TYPE.MINUTES && step % 5 !== 0)
            return '';
        return step.toString();
    };
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object)
    ], WClockComponent.prototype, "userTime", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object)
    ], WClockComponent.prototype, "minTime", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object)
    ], WClockComponent.prototype, "maxTime", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], WClockComponent.prototype, "userTimeChange", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Number)
    ], WClockComponent.prototype, "currentView", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", Object)
    ], WClockComponent.prototype, "viewChange", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], WClockComponent.prototype, "color", void 0);
    WClockComponent = __decorate([
        core_1.Component({
            selector: 'w-clock',
            styleUrls: ['./w-clock.component.scss'],
            templateUrl: './w-clock.component.html'
        })
    ], WClockComponent);
    return WClockComponent;
}());
exports.WClockComponent = WClockComponent;
